conan/tools/gnu/autotoolsdeps.py

```diff
@@ -21,43 +31,84 @@ def cpp_info(self):
+        if self._cpp_flags is None:
+            self._cpp_flags = []
+            self._cpp_flags.extend(self._flags.include_paths)
+            self._cpp_flags.extend(self._flags.defines)
+        return self._cpp_flags
+
+    @property
+    def cxx_flags(self):
+        self._check_env_already_calculated()
+        if self._cxx_flags is None:
+            self._cxx_flags = self._flags.cxxflags
+            if self._flags.sysroot:
+                self._cxx_flags.append(self._flags.sysroot)
+        return self._cxx_flags
+
+    @property
+    def c_flags(self):
+        self._check_env_already_calculated()
+        if self._c_flags is None:
+            self._c_flags = self._flags.cflags
+            if self._flags.sysroot:
+                self._c_flags.append(self._flags.sysroot)
+        return self._c_flags
+
+    @property
+    def ld_flags(self):
+        self._check_env_already_calculated()
+        if self._ld_flags is None:
+            self._ld_flags = self._flags.sharedlinkflags
+            self._ld_flags.extend(self._flags.exelinkflags)
+            self._ld_flags.extend(self._flags.frameworks)
+            self._ld_flags.extend(self._flags.framework_paths)
+            self._ld_flags.extend(self._flags.lib_paths)
+            # FIXME: Previously we had an argument "include_rpath_flags" defaulted to False
+            self._ld_flags.extend(self._flags.rpath_flags)
+            if self._flags.sysroot:
+                self._ld_flags.append(self._flags.sysroot)
+        return self._ld_flags
+
+    @property
     def environment(self):
-        flags = GnuDepsFlags(self._conanfile, self.cpp_info)
-
-        # cpp_flags
-        cpp_flags = []
-        cpp_flags.extend(flags.include_paths)
-        cpp_flags.extend(flags.defines)
-
-        # Ldflags
-        ldflags = flags.sharedlinkflags
-        ldflags.extend(flags.exelinkflags)
-        ldflags.extend(flags.frameworks)
-        ldflags.extend(flags.framework_paths)
-        ldflags.extend(flags.lib_paths)
-        # FIXME: Previously we had an argument "include_rpath_flags" defaulted to False
-        ldflags.extend(flags.rpath_flags)
-
-        # cflags
-        cflags = flags.cflags
-        cxxflags = flags.cxxflags
-
-        srf = flags.sysroot
-        if srf:
-            cflags.append(srf)
-            cxxflags.append(srf)
-            ldflags.append(srf)
-
-        env = Environment(self._conanfile)
-        env.append("CPPFLAGS", cpp_flags)
-        env.append("LIBS", flags.libs)
-        env.append("LDFLAGS", ldflags)
-        env.append("CXXFLAGS", cxxflags)
-        env.append("CFLAGS", cflags)
-        return env
-
-    def generate(self, env=None, auto_activate=True):
-        env = env or self.environment()
-        env.save_script("conanautotoolsdeps", auto_activate=auto_activate)
+        if not self._env:
+            env = Environment(self._conanfile)
+            env.append("CPPFLAGS", self._cpp_flags or self.cpp_flags)
```

Comment:
This idiom here ``self._cpp_flags or self.cpp_flags`` with the second also computing the former is very confusing.

Reply:
Agree. Needed tricks to allow that level of customization. I went to the maximum level of customization so we can decide the level we want.